### ISSUE TITLE
Define encoding of "password" to UTF-8

### ISSUE THREAD

**Issue by zonyitoo** [shadowsocks/shadowsocks-org#169]
## Motivation

The `"password"` could be passed in various ways, basic configration file, SIP008 online configuration, command line options, and other customized ways. Currently, the encoding of `"password"` is defined by implementations, for example

- `shadowsocks-libev`: UTF-8 in basic configuration file, but undefined in command line options
- `shadowsocks-rust`: UTF-8

So it is possible that if user uses `libev` with command line options and specifies `"password"` in UTF-16BE, it shouldn't work with the `rust` version, because the `"password"`s are not the same in byte level.

## Proposal

1. The `"password"` is a "string" in UTF-8 encoding
2. Implementations are allowed to accept `"password"` in other encoding, but they have to be converted it to UTF-8 before being set passing to ciphers

**Comment by Mygod** [comment]
You described the problem, not the motivation. Why do we need fancy characters in password anyway? A mix of upper/lower-case letters and numbers should always work, in which case the encoding does not matter.

**Comment by zonyitoo** [comment]
You are right. I am totally ok for limiting `"password"` to only ASCII characters, but the problem is that there is no standard about it. Implementations allow users to set non-ASCII `"password"` and that would eventually results in non-compatible between implementations.

**Comment by Mygod** [comment]
Sure but we should probably not include anything beyond ASCII.

**Comment by zonyitoo** [comment]
So we could make a conclusion:

1. `"password"` must be an UTF-8 string
2. `"password"` is recommended to contain only ASCII characters for maximum compatiblity

**Comment by dev4u** [comment]
如果能支持ascii外的可见字符，可以使密码不用设置得那么长，也可以得到较高的安全等级。
